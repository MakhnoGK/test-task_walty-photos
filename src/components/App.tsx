import React from 'react';
import { Container, Box, Typography } from '@material-ui/core';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import { ImageList, SearchForm } from '.';

import { ImageState } from '../types/imageState';

const App: React.FC<ImageState> = ({ data, loading }) => (
  <Container className="mb-5">
    <Typography variant="h3">Walty - graphics for all your needs</Typography>

    <Box>
      <SearchForm />
    </Box>

    <Box className="mb-5">
      {loading && (
        <Typography className="text-center mt-2">Loading...</Typography>
      )}

      {data.total > 0 && (
        <Typography className="text-right mt-2">
          Results: {data.totalHits} of {data.total}
        </Typography>
      )}

      <ImageList images={data.hits} />
    </Box>

    <Box>
      <Box className="d-flex align-items-center justify-content-center">
        Powered by{' '}
        <a
          href="http://pixabay.com/"
          rel="noreferrer"
          target="_blank"
          className="ml-2"
        >
          <img
            src="https://pixabay.com/static/img/logo.svg"
            alt="Pixabay"
            width="150"
          />
        </a>
      </Box>
    </Box>
  </Container>
);

App.propTypes = {
  data: PropTypes.shape({
    hits: PropTypes.array.isRequired,
    total: PropTypes.number.isRequired,
    totalHits: PropTypes.number.isRequired,
  }).isRequired,
  loading: PropTypes.bool.isRequired,
};

const propsMap = (state: ImageState) => ({
  data: state.data,
  loading: state.loading,
});

export default connect(propsMap)(App);
