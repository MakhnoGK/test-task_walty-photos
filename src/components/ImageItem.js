import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

const ImageContainer = styled.div`
  flex: 1 1 300px;
  width: 100%;
  height: 300px;
  position: relative;
  display: flex;
  justify-content: space-between;
  align-items: flex-end;

  &::after {
    content: "";
    z-index:-2;
    position: absolute;
    width: 80%;
    height: 150px;
    bottom: 0;
    left: 50%;
    transform: translateX(-50%);
    box-shadow: 0 5px 10px rgb(0 0 0 / 40%);  
  }
`;

const Image = styled.img`
  position: absolute;
  width: 100%;
  height: 100%;
  top: 0;
  right: 0;
  bottom: 0;
  left: 0;
  object-fit: cover;
  z-index: -1;
`;

const Info = styled.div`
  color: #e5e5e5;
  margin: 0.5rem;
  background: #00000088;
  padding: 5px;
`;

const TagLink = styled.a`
display:block;
  text-decoration: none;
  color: inherit;
  text-align: right;

  &:hover {
    color: currentColor;
  }
`

const ImageItem = (props) => {
  const { type, tags, webformatURL } = props.image;

  const displayTags = (input) => {
    return input.split(', ').map((tag) => <TagLink href={`/${tag}`}>{tag}</TagLink>);
  };

  return (
    <ImageContainer>
      <Info>{type}</Info>
      <Info>{displayTags(tags)}</Info>
      <Image alt={tags} src={webformatURL} />
    </ImageContainer>
  );
};

ImageItem.propTypes = {
  image: PropTypes.shape({
    webformatURL: PropTypes.string,
    largeImageURL: PropTypes.string,
    type: PropTypes.string,
    tags: PropTypes.string,
  }),
};

export default ImageItem;
