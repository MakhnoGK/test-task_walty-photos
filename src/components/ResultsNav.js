import React from 'react';
import PropTypes from 'prop-types';
import { Row } from 'react-bootstrap';

const ResultsNav = ({ total }) => {
  return (
    <Row className="mb-3 align-items-center">
      <a href="/" className="btn btn-light d-inline-block mr-2">
        &laquo; Back
      </a>

      <p className="m-0">Total results: {total}</p>
    </Row>
  );
};

ResultsNav.propTypes = {
  total: PropTypes.number,
};

export default ResultsNav;
