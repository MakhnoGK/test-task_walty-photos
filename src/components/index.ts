export { default as App } from './App';
export { default as ImageItem } from './image-list/ImageItem';
export { default as ImageList } from './image-list/ImageList';
export { default as SearchForm } from './image-list/SearchForm';
export { default as ImageTags } from './image-list/ImageTags';
