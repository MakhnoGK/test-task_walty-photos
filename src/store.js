import createSagaMiddleware from 'redux-saga';
import { applyMiddleware, createStore } from 'redux';
import { composeWithDevTools } from 'redux-devtools-extension';

import imagesReducer from './ducks/images';
import watchFetchImages from './sagas/imagesSaga';

const sagaMiddleware = createSagaMiddleware();
const appStore = createStore(
  imagesReducer,
  composeWithDevTools(applyMiddleware(sagaMiddleware))
);

sagaMiddleware.run(watchFetchImages);

export default appStore;
