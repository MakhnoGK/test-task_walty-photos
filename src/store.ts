import createSagaMiddleware from 'redux-saga';
import { applyMiddleware, createStore } from 'redux';
import { composeWithDevTools } from 'redux-devtools-extension';

import imagesReducer, { watchFetchImages } from './ducks/images';

const sagaMiddleware = createSagaMiddleware();

const store = createStore(
  imagesReducer,
  composeWithDevTools(applyMiddleware(sagaMiddleware))
);

sagaMiddleware.run(watchFetchImages);

export default store;
