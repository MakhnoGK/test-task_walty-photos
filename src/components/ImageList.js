import React from 'react';
import PropTypes from 'prop-types';
import ImageItem from './ImageItem';
import styled from 'styled-components';

const ListContainer = styled.div`
  display: flex;
  flex-wrap: wrap;
  gap: 2rem;
`;

const ImageList = (props) => {
  const { images } = props;

  return (
    <ListContainer>
      {images.map((image) => (
        <ImageItem image={image} key={image.id} />
      ))}
    </ListContainer>
  );
};

ImageList.propTypes = {
  images: PropTypes.array,
};

export default ImageList;
