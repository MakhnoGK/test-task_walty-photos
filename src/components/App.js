import React, { useState} from 'react';
import { Container, Row, Col } from 'react-bootstrap';
import ImageList from './ImageList';
import SearchForm from './SearchForm';
import { useTags } from '../hooks/useTags';

const API_KEY = '20295782-d190a9f4db1bc0031bd7c6307';

const App = () => {
  const [tags, setTags] = useTags([]);
  const [loading, setLoading] = useState(false);
  const [imagesData, setImagesData] = useState({
    images: [],
    total: 0,
    totalHits: 0,
  });

  const fetchImageData = async () => {
    const term = tags.map((tag) => tag.replaceAll(/\W+/g, '')).join('+');

    return fetch(`https://pixabay.com/api/?key=${API_KEY}&q=${term}`)
      .then((response) => response.json())
      .then((data) => data);
  };

  const handleTags = (input) => {
    setTags(input.values);
  };

  const handleSearch = () => {
    setLoading(true);

    fetchImageData().then((imageData) => {
      setLoading(false);
      setImagesData((state) => ({
        ...state,
        images: imageData.hits,
        total: imageData.total,
        totalHits: imageData.totalHits,
      }));
    });
  };

  return (
    <Container className="mb-5">
      <h2 className="text-center my-4">Walty - graphics for all your needs</h2>

      <Row className="mb-4">
        <Col>
          <SearchForm onSearch={handleSearch} onTags={handleTags} tags={tags} />
        </Col>
      </Row>

      <Row className="mb-5">
        <Col>
          {loading && <p className="text-center mt-2">Loading...</p>}

          {imagesData.total > 0 && (
            <p className="text-right mt-2">
              Results: {imagesData.totalHits} of {imagesData.total}
            </p>
          )}
          <ImageList images={imagesData.images} />
        </Col>
      </Row>

      <Row>
        <Col>
          <div className="d-flex align-items-center justify-content-center">
            Powered by{' '}
            <a
              href="http://pixabay.com/"
              rel="noreferrer"
              target="_blank"
              className="ml-2"
            >
              <img
                src="https://pixabay.com/static/img/logo.svg"
                alt="Pixabay"
                width="150"
              />
            </a>
          </div>
        </Col>
      </Row>
    </Container>
  );
};

App.propTypes = {};

export default App;
