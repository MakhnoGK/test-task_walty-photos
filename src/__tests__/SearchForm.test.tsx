import React from 'react';
import { Provider } from 'react-redux';
import { MemoryRouter } from 'react-router-dom';
import renderer from 'react-test-renderer';

import { SearchForm } from '../components';

import store from '../redux';

describe('SearchForm tests', () => {
  it('renders correctly', () => {
    const tree = renderer
      .create(
        <Provider store={store}>
          <MemoryRouter initialEntries={['/tag1']}>
            <SearchForm />
          </MemoryRouter>
        </Provider>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
