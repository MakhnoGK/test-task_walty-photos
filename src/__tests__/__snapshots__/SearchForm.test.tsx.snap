// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchForm tests renders correctly 1`] = `
<div
  className="input-group"
>
  <div
    className="form-control d-inline-flex flex-wrap"
  >
    <input
      aria-label="photo, yellow, flower"
      autoFocus={true}
      className="border-0 w-auto flex-fill input-tags"
      data-testid="input-tags"
      onChange={[Function]}
      onKeyUp={[Function]}
      placeholder="photo, yellow, flower"
      type="text"
      value=""
    />
  </div>
</div>
`;
